File: kernels/Python/compositron/examples/_bench_cdbs.py
from time import perf_counter_ns
import numpy as np

from compositron.cdbs.cdbspectrum import Axis, DiagonalArea, LinearProjectionBins, LineshapeParamDefinition
from compositron.core.measurement import Measurement
from compositron.core.utils import EcalCorrectionOrder, KeV


def print_mean_and_std(name: str, data: np.ndarray) -> None:
    mean = np.mean(data)
    std = np.std(data)

    print(f"{name}: {mean} ns +/- {std} ns")


def main():
    n_iter = 1000

    imports = np.empty(n_iter)
    corr = np.empty(n_iter)
    s_calc = np.empty(n_iter)
    proj_axes = np.empty(n_iter)
    proj_diag = np.empty(n_iter)

    for i in range(n_iter):
        start = perf_counter_ns()
        m = Measurement(
            "../../../../testdata/depth-profile_Copper_0000.n42"
        )
        stop = perf_counter_ns()
        imports[i] = stop - start

        c = m.cdbs["OAA x OAB"]

        start = perf_counter_ns()
        c.correct_ecal(EcalCorrectionOrder.FIRST)
        stop = perf_counter_ns()
        corr[i] = stop - start

        param = LineshapeParamDefinition(
            name = "S",
            num = [DiagonalArea(KeV(2), KeV(1))],
            denom = [DiagonalArea(KeV(10), KeV(1))],
            in_corrected_peak = False,
        )

        start = perf_counter_ns()
        c.calc_lineshape_param(param)
        stop = perf_counter_ns()
        s_calc[i] = stop - start

        start = perf_counter_ns()
        c.project_digonal(
            LinearProjectionBins(KeV(0.1)),
            KeV(2.),
            KeV(100.),
            False
        )
        stop = perf_counter_ns()
        proj_diag[i] = stop - start

        start = perf_counter_ns()
        c.project_axes(Axis.FIRST_DET)
        stop = perf_counter_ns()
        proj_axes[i] = stop - start

    print_mean_and_std("Import", imports)
    print_mean_and_std("Ecal Corr", corr)
    print_mean_and_std("S Param", s_calc)
    print_mean_and_std("Proj Diag", proj_diag)
    print_mean_and_std("Proj Axes", proj_axes)


if __name__ == "__main__":
    main()

File: kernels/Rust/compositron/examples/_bench_cdbs.rs
use std::time::Instant;

use compositron::core::Measurement;
use compositron::core::utils::{Unit, EcalCorrectionOrder};
use compositron::cdbs::cdbspectrum::{
    Area, Axis, LineshapeParamDefinition, ProjectionBins
};
use compositron::importers::DataFormat;


fn print_mean_and_std(name: &str, data: &Vec<u128>) {
    let n = data.len();
    let mean = data.iter().sum::<u128>() as f64 / n as f64;
    let std = (
        data
            .iter()
            .map(|&x| (x as f64 - mean).powi(2))
            .sum::<f64>() / n as f64
    ).sqrt();

    println!("{name}: {mean} ns +/- {std} ns");
}


fn main() -> anyhow::Result<()> {
    let n_iter = 10000;

    let mut imports = Vec::with_capacity(n_iter);
    let mut corr = Vec::with_capacity(n_iter);
    let mut s_calc = Vec::with_capacity(n_iter);
    let mut proj_axes = Vec::with_capacity(n_iter);
    let mut proj_diag = Vec::with_capacity(n_iter);

    for _ in 0..n_iter {
        let start = Instant::now();
        let mut m = Measurement::from_file(
            "../../../../testdata/depth-profile_Copper_0000.n42",
            DataFormat::SlopeN42
        )?;
        imports.push(start.elapsed().as_nanos());

        let c = m.cdbs.get_mut("OAA x OAB").unwrap();

        let start = Instant::now();
        c.correct_ecal(EcalCorrectionOrder::First)?;
        corr.push(start.elapsed().as_nanos());

        let ls_param = LineshapeParamDefinition {
            name: "S",
            num: &[
                Area::Diagonal {
                    width_cel: Unit::KeV(2.),
                    width_cml: Unit::KeV(1.),
                },
            ],
            denom: &[
                Area::Diagonal {
                    width_cel: Unit::KeV(10.),
                    width_cml: Unit::KeV(1.),
                },
            ],
            in_corrected_peak: false,
        };

        let start = Instant::now();
        let s = c.calc_lineshape_param(&ls_param)?;
        s_calc.push(start.elapsed().as_nanos());

        let start = Instant::now();
        let p = c.project_digonal(
            ProjectionBins::Linear(Unit::KeV(0.1)),
            Unit::KeV(2.),
            Unit::KeV(100.),
            false
        )?;
        proj_diag.push(start.elapsed().as_nanos());

        let start = Instant::now();
        let p = c.project_axes(Axis::FirstDetector);
        proj_axes.push(start.elapsed().as_nanos());
    }

    print_mean_and_std("Import", &imports);
    print_mean_and_std("Ecal Corr", &corr);
    print_mean_and_std("S Param", &s_calc);
    print_mean_and_std("Proj Diag", &proj_diag);
    print_mean_and_std("Proj Axes", &proj_axes);

    Ok(())
}
